src/reader_disc.cpp

```diff
@@ -0,0 +1,559 @@
+    }
+
+    if (p_sector[0] == 0x00 && p_sector[1] == 0x00 && p_sector[2] == 0x01 &&
+        p_sector[3] == 0xba) {
+      goto nocache;
+    }
+
+    /* The data we are looking for is at sector 16 (32768 bytes):
+     *  - offset 40: disc title (32 uppercase chars)
+     *  - offset 813: manufacturing date + serial no (16 digits) */
+
+    if (!read(dev, p_sector, 16, 1)) {
+      goto nocache;
+    }
+
+    /* Get the disc title */
+    psz_title = (char *)p_sector + 40;
+    psz_title[32] = '\0';
+
+    for (i = 0; i < 32; i++) {
+      if (psz_title[i] <= ' ') {
+        psz_title[i] = '\0';
+        break;
+      } else if (psz_title[i] == '/' || psz_title[i] == '\\') {
+        psz_title[i] = '-';
+      }
+    }
+
+    /* Get the date + serial */
+    psz_serial = p_sector + 813;
+    psz_serial[16] = '\0';
+
+    /* Check that all characters are digits, otherwise convert. */
+    for (i = 0; i < 16; i++) {
+      if (psz_serial[i] < '0' || psz_serial[i] > '9') {
+        char psz_tmp[16 + 1];
+        sprintf(psz_tmp, "%.2x%.2x%.2x%.2x%.2x%.2x%.2x%.2x", psz_serial[0],
+                psz_serial[1], psz_serial[2], psz_serial[3], psz_serial[4],
+                psz_serial[5], psz_serial[6], psz_serial[7]);
+        memcpy(psz_serial, psz_tmp, 16);
+        break;
+      }
+    }
+
+    /* Get disk key, since some discs have got same title, manufacturing
+     * date and serial number, but different keys */
+    if (dev->media.dvdcss.protection == 0x01) {
+      psz_key[0] = '-';
+      for (i = 0; i < DVD_KEY_SIZE; i++) {
+        sprintf(&psz_key[1 + i * 2], "%.2x", dev->media.dvdcss.DK[i]);
+      }
+      psz_key[1 + DVD_KEY_SIZE * 2] = '\0';
+    } else {
+      psz_key[0] = 0;
+    }
+
+    /* We have a disc name or ID, we can create the cache dir */
+    i = sprintf(dev->media.dvdcss.psz_cachefile, "%s", psz_cache);
+    if (mkdir(dev->media.dvdcss.psz_cachefile, 0755) < 0 && errno != EEXIST) {
+      printf("CSS: failed creating cache directory\n");
+      dev->media.dvdcss.psz_cachefile[0] = '\0';
+      goto nocache;
+    }
+
+    i += sprintf(dev->media.dvdcss.psz_cachefile + i, "/%s-%s%s", psz_title,
+                 psz_serial, psz_key);
+    if (mkdir(dev->media.dvdcss.psz_cachefile, 0755) < 0 && errno != EEXIST) {
+      printf("CSS: failed creating cache subdirectory\n");
+      dev->media.dvdcss.psz_cachefile[0] = '\0';
+      goto nocache;
+    }
+    i += sprintf(dev->media.dvdcss.psz_cachefile + i, "/");
+
+    /* Pointer to the filename we will use. */
+    dev->media.dvdcss.psz_block = dev->media.dvdcss.psz_cachefile + i;
+
+    printf("CSS: using CSS key cache dir: %s\n",
+           dev->media.dvdcss.psz_cachefile);
+  }
+nocache:
+  return 0;
+}
+#endif
+
+/* Loop over all titles and call dvdcss_title to crack the keys. */
+static int initAllCSSKeys(drive_info *dev) {
+  struct timeval all_s, all_e;
+  struct timeval t_s, t_e;
+  char filename[MAX_UDF_FILE_NAME_LEN];
+  unsigned int start, len;
```

Comment:
Codacy found an issue: [The scope of the variable 'start' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)